appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockChart.java

```diff
@@ -0,0 +1,201 @@
+        setTypeProperty("0");
+
+        initComponent(rootPanel);
+
+        // Re-attach all children MockChartData components
+        rootPanel.addAttachHandler(new AttachEvent.Handler() {
+            @Override
+            public void onAttachOrDetach(AttachEvent arg0) {
+                if (arg0.isAttached()) {
+                    for (MockComponent child : children) {
+                        ((MockChartData) child).addToChart(MockChart.this);
+                    }
+                }
+            }
+        });
+    }
+
+    /**
+     * Sets the type of the Chart to the newly specified value.
+     * @param value  new Chart type
+     */
+    private void setTypeProperty(String value) {
+        // Update type
+        type = Integer.parseInt(value);
+
+        // Keep track whether this is the first time that
+        // the Chart view is being initialized
+        boolean chartViewExists = (chartView != null);
+
+        // Remove the current Chart Widget from the root panel (if present)
+        if (chartViewExists) {
+            rootPanel.remove(chartView.getChartWidget());
+        }
+
+        // Create a new Chart view based on the supplied type
+        chartView = createMockChartViewFromType(type);
+
+        // Add the Chart Widget to the Root Panel (as the first widget)
+        rootPanel.insert(chartView.getChartWidget(), 0);
+
+        // Chart view already existed before, so the new Chart view must
+        // be reinitialized.
+        if (chartViewExists) {
+            reinitializeChart();
+        }
+    }
+
+    /**
+     * Creates and returns a new MockChartView object based on the type
+     * (integer) provided
+     * @param type  Chart type (integer representation)
+     * @return new MockChartView object instance
+     */
+    private MockChartView createMockChartViewFromType(int type) {
+        switch(type) {
+            case 0:
+                // Line Chart
+                return new MockLineChartView();
+            case 1:
+                // Scatter Chart
+                return new MockLineChartView();
+            case 2:
+                // Area Chart
+                return new MockLineChartView();
+            case 3:
+                // Bar Chart
+                return new MockLineChartView();
+            case 4:
+                // Pie Chart
+                return new MockLineChartView();
+            default:
+                // Invalid argument
+                throw new IllegalArgumentException("type:" + type);
+        }
+    }
+
+    /**
+     * Reinitializes the Chart view by reattaching all the Data
+     * components and setting back all the properties.
+     */
+    private void reinitializeChart() {
+        // Chart type changing requires setting back Chart-related properties
+        chartView.setBackgroundColor(getPropertyValue(PROPERTY_NAME_BACKGROUNDCOLOR));
+        chartView.setTitle(getPropertyValue(PROPERTY_NAME_DESCRIPTION));
+        chartView.getChartWidget().draw();
```

Comment:
Is there any better, more consistent way of re-setting the properties?

The reason this is needed is because the Chart view gets reinitialized dynamically, so the properties have to be set for the new view as well.

See line 88 (the Type setter) for the context of this method.